File: src/Score.hs
{-# LANGUAGE ConstraintKinds, FlexibleContexts, RankNTypes, OverloadedStrings #-}
{-# LANGUAGE UndecidableInstances, ScopedTypeVariables, AllowAmbiguousTypes   #-}
-----------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------
-- | 
-- | Module  : All functions used to find score
-- | Author  : Xiao Ling
-- | Date    : 9/14/2016
-- |             
---------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------

module Score (
    sumCnt
  , cnt


  ) where

import System.FilePath.Posix
import Control.Monad.Trans.Reader
import Data.Attoparsec.Text
import Data.Attoparsec.Combinator
import Data.Text (Text, unpack, pack)

import Lib
import Core
import Query

{-----------------------------------------------------------------------------
  Score 
------------------------------------------------------------------------------}

w1 :: String -> String -> ReaderT Config IO Output
w1 a1 a2 = do
  path_ws <- weakStrong <$> ask 
  

  --p_ws <- pattern weakStrong
  --sumCount $ (\p -> p (S a1) (S a2)) <$> p_ws

{-
s1 :: String -> String -> ReaderT Config IO Output
s1 a1 a2 = do
  p_sw <- pattern strongWeak
  sumCount $ (\p -> p (S a1) (S a2)) <$> p_sw

w2 :: String -> String -> ReaderT Config IO Output
w2 = flip w1

s2 :: String -> String -> ReaderT Config IO Output
s2 = flip s1

p1 :: ReaderT Config IO Output
p1 = do
  p_ws <- pattern weakStrong
  sumCount $ (\p -> p Star Star) <$> p_ws

p2 :: ReaderT Config IO Output
p2 = do
  p_sw <- pattern strongWeak
  sumCount $ (\p -> p Star Star) <$> p_sw
-}


{-----------------------------------------------------------------------------
  Count
------------------------------------------------------------------------------}

-- * sum the results of multiple `count`s 
-- * and sum their counts, list all results
sumCnt :: [Parser Text] -> ReaderT Config IO Output
sumCnt ps = do
  rrs <- mapM cnt ps
  let ns = fst <$> rrs
  let rs = snd <$> rrs
  return (sum ns, concat rs)

cnt :: Parser Text -> ReaderT Config IO Output
cnt p = do
  d  <- dataDir <$> ask
  os <- query (pathToData d p) p
  return os

{-----------------------------------------------------------------------------
  File System Utils
------------------------------------------------------------------------------}

-- * @Use : pathToData "path/to/data" "path/to/pattern.txt" someParser
-- * Given path to data directory `d`, and parser `p`
-- * Build path to data to be queried by parser `p`
-- * Note  : This is *tightly coupled* to naming conventions of this project
pathToData :: DirectoryPath -> Parser Text -> FilePath
pathToData d p = takeDirectory d ++ "/" ++ name p ++ ".txt"


{-----------------------------------------------------------------------------
  Score 

w1 :: String -> String -> ReaderT Config IO Output
w1 a1 a2 = do
  p_ws <- pattern weakStrong
  sumCount $ (\p -> p (S a1) (S a2)) <$> p_ws

s1 :: String -> String -> ReaderT Config IO Output
s1 a1 a2 = do
  p_sw <- pattern strongWeak
  sumCount $ (\p -> p (S a1) (S a2)) <$> p_sw

w2 :: String -> String -> ReaderT Config IO Output
w2 = flip w1

s2 :: String -> String -> ReaderT Config IO Output
s2 = flip s1

p1 :: ReaderT Config IO Output
p1 = do
  p_ws <- pattern weakStrong
  sumCount $ (\p -> p Star Star) <$> p_ws

p2 :: ReaderT Config IO Output
p2 = do
  p_sw <- pattern strongWeak
  sumCount $ (\p -> p Star Star) <$> p_sw

{-----------------------------------------------------------------------------
  Count
------------------------------------------------------------------------------}

-- * sum the results of multiple `count`s 
-- * and sum their counts, list all results
sumCount :: [Parser Text] -> ReaderT Config IO Output
sumCount ps = do
  rrs <- mapM countp ps
  let ns = fst <$> rrs
  let rs = snd <$> rrs
  return (sum ns, concat rs)

-- * `count` for occurences of some phrase among ngram files
countp :: Parser Text -> ReaderT Config IO Output
countp phrase = do
  con     <- ask
  (n, ts) <- phrase `query` (ngrams con)
  return (n,ts)

-- * `count` occurences of some word `w` 
-- * in onegram file
count :: String -> ReaderT Config IO Output
count w = do
  let word = compile' w 
  con     <- ask
  (n, ts) <- word `query` [onegram con]
  return (n,ts)

------------------------------------------------------------------------------}

File: python/score.py
############################################################
# Module  : Scores functio
# Date    : Oct. 30th
# Author  : Xiao Ling
############################################################


############################################################
# Score Function
############################################################

# @Use    : Given DataReader Object and a list of words `ws`, 
# @Output : score(ai,aj) for each i j in ws, 
#           normalized by min score

# scores :: DataReader -> [String] -> Error String (Dict String Float)
def scores(O,ws):
	pws  = [(u,v) for u in ws for v in ws]
	d    = dict()
	minv = 1e10

	for (u,v) in pws:
		n = toScore(O,u,v)
		if n and abs(n) < minv: minv = abs(n)
		d[u + "_" + v] = n

	for uv in d:
		d[uv] = d[uv]/minv

	return d

# @Use   : Given DataReader Object and words ai aj, output:
# @Output: score(ai,ak) where:
# 
#          score(ai,ak) =   (W1 - S1) - (W2 - S2)
#                         --------------------------
#                           cnt(ai) * cnt(ak)
# 
#          where W1 = 1/P1 * w1 = 1/P1 * sum_{p in P_ws} cnt(p(ai,ak))
#          where W2 = 1/P1 * w2 = 1/P1 * sum_{p in P_ws} cnt(p(ak,ai))
#          where S1 = 1/P2 * s1 = 1/P2 * sum_{p in P_sw} cnt(p(ai,ak))
#          where W2 = 1/P2 * s2 = 1/P2 * sum_{p in P_sw} cnt(p(ak,ai))

# toScore :: DataReader -> String -> String -> Error String Float
def toScore(O,ai,ak):
	ws = [ai,ak]
	d  = O.total()
	P1 = d['weak-strong-normalization']
	P2 = d['strong-weak-normalization']

	return goScores( O.count (ws) \
		           , O.strong(ws) \
		           , O.weak  (ws) \
		           , P1
		           , P2
		           , ai
		           , ak)

# goScores :: Dict String Float
#        -> Dict String Float
#        -> Dict String Float
#        -> Dict String Float
#        -> Float
#        -> Float
#        -> String
#        -> String
#        -> Error String Float
def goScores(count,strong,weak,P1,P2,ai,ak):

	if ai in count and ak in count:
		w1 = sumcnt(weak   , ai , ak)
		w2 = sumcnt(weak   , ak , ai)
		s1 = sumcnt(strong , ai , ak)
		s2 = sumcnt(strong , ak , ai)

		top = (w1/P1 - s1/P2) - (w2/P1 - s2/P2)
		bot = count[ai] * count[ak]

		return (top/bot)

	else:
		raise NameError("words: " + ai + " and " + ak + "not found")


# @Use: Given a `scores` dictonary mapping word pairs to 
#       their weak and strong scores
#       and words `u` `v`, output their total occurences

# sumcnt :: Dict String Float -> String -> String -> Error String Float
def sumcnt(scores, u, v):
	name = u + "-" + v
	if name in scores:
		return sum(scores[name])
	else:
		raise NameError("key " + name + " not found")





